ui/lib/client/index.tsx

```diff
@@ -29,14 +60,36 @@ function initAxios() {
       message.error({ content, key: 'network_error' }) // use the same key to avoid multiple message boxes
       err.handled = true
       err.msg = content // use `err.message = content` doesn't work
+    } else if (handleErrorWay !== 'custom') {
+      const fullUrl = config.url as string
+      const API = fullUrl.replace(getBasePath(), '').split('?')[0]
+      const errCode = response?.data?.code || 'error.message.unknown'
+      const content = i18next.t(errCode)
+
+      if (handleErrorWay === 'notification') {
+        notification.error({
```

Comment:
But modal will block the user's operation, notification or message will auto close after a duration.

Reply:
Yes, blocking the user operation is the intentional purpose. Currently there are these places handling errors:

- SignIn: Using a custom handler, display error message inline
- Most GET requests via useClientRequest hook: It will grab errors in `error` property and which will then be displayed by an error bar.
- The rest of the GET requests: you have already refined it to use a custom handler
- The rest: looks like the remaining is all operational requests, like POST, PUT or DELETE. Currently some of them are manually `catch`ed and when there are errors a Modal is displayed. Handling error globally will unify this behaviour to cover all operational requests.

Reply:
Still have many GET requests we don't handle it with a custom handler, for example this:

```js
  const { data: tableData, isLoading } = useClientRequest((cancelToken) =>
    client.getInstance().getHostsInfo({ cancelToken })
  )
```

Is it fine that inside the default handler, we use notification for GET request while use modal for POST/PUT/DELETE request?

Reply:
It's a good idea! There are something difficult though. useClientRequest may or may not handle error, depend on its caller. How can we know that the notification should be shown or not?

Reply:
The logic is this: when sending a request, we can set the errorStrategy to the request config. If we want to handle the error by the component self, then we can set its value to `Custom`, else set it to `Default` or don't set it.

```js
  const { data } = useClientRequest((cancelToken) =>
    client.getInstance().infoGet({ cancelToken, errorStrategy: ErrorStrategy.Custom })
  )
```

If this API meets error, it will enter the axios response interceptor first, inside the interceptor, we can retrieve the original errorStrategy from the request config. We judge its value, if the value is `Custom`, then we don't deal with it, else if its value is `Default`, then we show the error message by notification or modal.

Whether we handle the error inside the interceptor, we will throw the error up. (so although the interceptor handle the error, the component can still handle it again if it would like).

Not sure whether I explain it clearly.

Reply:
I got it. Maybe we can modify `useClientRequest` to pass a `options: { cancelToken, errorStrategy }` by default directly, so that we can make callers shorter, no need to write `{ cancelToken, errorStrategy: ErrorStrategy.Custom }` every time. Instead, just passing an `options`.

Reply:
Ok, let me have a try.

Reply:
Update the useClientRequest to:

```
      const reqConfig: ReqConfig = {
        cancelToken: cancelTokenSource.current.token,
        errorStrategy: handleError
          ? ErrorStrategy.Default
          : ErrorStrategy.Custom,
      }
      const resp = await reqFactory(reqConfig)
```

Now the caller like this:

```
  const {
    data: dataStores,
    isLoading: loadingStores,
    error: errStores,
  } = useClientRequest(
    (reqConfig) => client.getInstance().getStoreTopology(reqConfig),
    { handleError: false }
  )
```

Because the errorStrategy only has 2 values, I am a little considering to change it to `handleError: boolean`.